File: data/data.code.javascript.client/协议模板/ReqTestSend.js
module.exports = ReqTestSend;

var Packet = require('../net/Packet');
var Msg = require('./Msg');
var MsgRoleBase = require('./MsgRoleBase');


var ReqTestSend = function() {
	this._id_u8 = 0;
	this._role_base = undefined;
	this._id_f32 = new Array();
	this._id_op_u8_flag = 0;
	this._id_op_u8 = 0;
	this._op_role_base_flag = 0;
	this._op_role_base = undefined;


	this.Encode() {
		var packet = new Packet();
		packet.WriteByte(this._id_u8);
		packet.WriteBuffer(this._role_base.GetBuffer());
		var id_f32_count = this._id_f32.length;
		packet.WriteUshort(id_f32_count);
		for (var i = 0; i < id_f32_count; i++) {
			var xxx = this._id_f32[i];
			packet.WriteFloat(xxx);
		}
		packet.WriteByte(this._id_op_u8_flag);
		if (this._id_op_u8_flag) {
			packet.WriteByte(this._id_op_u8);
		}
		packet.WriteByte(this._op_role_base_flag) {
			packet.WriteByffer(this._op_role_base.GetBuffer());
		}
		packet.Encode(Msg.P_REQ_TEST_SEND);
		return packet;
	}


	this.SetIdU8(val) {
		this._id_u8 = val;
	}

	this.SetRoleBase(val) {
		this._role_base = val;
	}

	this.SetIdF32(val) {
		this._id_f32 = val;
	}

	this.SetIdOpU8(val) {
		this._id_op_u8_flag = 1;
		this._id_op_u8 = val;
	}
	
	this.SetOpRoleBase(val) {
		this._op_role_base_flag = 1;
		this._op_role_base = val;
	}
}

File: examples/golang/server/proto/req_test_send.go
package proto

import (
	"packet"
)

type ReqTestSend struct {
	IdU8                     uint8
	RoleBase                 *MsgRoleBase
	IdF32                    []float32
	IdOpU8                   uint8
	OpRoleBase               *MsgRoleBase
}

func ReqTestSendDecode(pack *packet.Packet) *ReqTestSend {
	reqTestSend := &ReqTestSend{}

	reqTestSend.IdU8 = pack.ReadUint8()
	reqTestSend.RoleBase = MsgRoleBaseDecode(pack)
	IdF32Count := pack.ReadUint16()
	for ;IdF32Count > 0; IdF32Count-- {
		reqTestSend.IdF32 = append(reqTestSend.IdF32, pack.ReadFloat32())
	}
	IdOpU8Flag := pack.ReadUint8()
	if IdOpU8Flag == 1 {
		reqTestSend.IdOpU8 = pack.ReadUint8()
	}
	OpRoleBaseFlag := pack.ReadUint8()
	if OpRoleBaseFlag == 1 {
		reqTestSend.OpRoleBase = MsgRoleBaseDecode(pack)
	}
	return reqTestSend
}
